README
NAME

    Test::Ability

ABSTRACT

    Property-Based Testing for Perl 5

SYNOPSIS

      package main;
    
      use Test::Ability;
    
      my $t = Test::Ability->new;

DESCRIPTION

    This package provides methods for generating values and test-cases,
    prodiving a framework for performing property-based testing.

INTEGRATES

    This package integrates behaviors from:

    Data::Object::Role::Buildable

    Data::Object::Role::Stashable

LIBRARIES

    This package uses type constraints from:

    Types::Standard

SCENARIOS

    This package supports the following scenarios:

 stash

      # given: synopsis
    
      $t->stash(direction => sub {
        my ($self) = @_;
    
        {
          move => ('forward', 'reverse')[rand(1)],
          time => time
        }
      });

    The package provides a stash object for default and user-defined value
    generators. You can easily extend the default generators by adding your
    own. Once defined, custom generators can be specified in the gen-spec
    (generator specification) arrayref provided to the test method (and
    others).

ATTRIBUTES

    This package has the following attributes:

 arguments

      arguments(ArrayRef)

    This attribute is read-only, accepts (ArrayRef) values, and is
    optional.

 invocant

      invocant(Object)

    This attribute is read-only, accepts (Object) values, and is optional.

METHODS

    This package implements the following methods:

 array

      array(Maybe[Int] $min, Maybe[Int] $max) : ArrayRef

    The array method returns a random array reference.

    array example #1

        # given: synopsis
      
        $t->array;

 array_object

      array_object(Maybe[Int] $min, Maybe[Int] $max) : Object

    The array_object method returns a random array object.

    array_object example #1

        # given: synopsis
      
        $t->array_object;

 choose

      choose(ArrayRef[ArrayRef] $args) : Any

    The choose method returns a random value from the set of specified
    generators.

    choose example #1

        # given: synopsis
      
        $t->choose([['datetime'], ['words', [2,3]]]);

 code

      code(Maybe[Int] $min, Maybe[Int] $max) : CodeRef

    The code method returns a random code reference.

    code example #1

        # given: synopsis
      
        $t->code;

 code_object

      code_object(Maybe[Int] $min, Maybe[Int] $max) : Object

    The code_object method returns a random code object.

    code_object example #1

        # given: synopsis
      
        $t->code_object;

 date

      date(Maybe[Str] $min, Maybe[Str] $max) : Str

    The date method returns a random date.

    date example #1

        # given: synopsis
      
        $t->date;

 datetime

      datetime(Maybe[Str] $min, Maybe[Str] $max) : Str

    The datetime method returns a random date and time.

    datetime example #1

        # given: synopsis
      
        $t->datetime;

 hash

      hash(Maybe[Int] $min, Maybe[Int] $max) : HashRef

    The hash method returns a random hash reference.

    hash example #1

        # given: synopsis
      
        $t->hash;

 hash_object

      hash_object(Maybe[Int] $min, Maybe[Int] $max) : Object

    The hash_object method returns a random hash object.

    hash_object example #1

        # given: synopsis
      
        $t->hash_object;

 maybe

      maybe(ArrayRef[ArrayRef] $args) : Any

    The maybe method returns a random choice using the choose method, or
    the undefined value.

    maybe example #1

        # given: synopsis
      
        $t->maybe([['date'], ['time']]);

 number

      number(Maybe[Int] $min, Maybe[Int] $max) : Int

    The number method returns a random number.

    number example #1

        # given: synopsis
      
        $t->number;

 number_object

      number_object(Maybe[Int] $min, Maybe[Int] $max) : Object

    The number_object method returns a random number object.

    number_object example #1

        # given: synopsis
      
        $t->number_object;

 object

      object() : Object

    The object method returns a random object.

    object example #1

        # given: synopsis
      
        $t->object;

 regexp

      regexp(Maybe[Str] $exp) : RegexpRef

    The regexp method returns a random regexp.

    regexp example #1

        # given: synopsis
      
        $t->regexp;

 regexp_object

      regexp_object(Maybe[Str] $exp) : Object

    The regexp_object method returns a random regexp object.

    regexp_object example #1

        # given: synopsis
      
        $t->regexp_object;

 scalar

      scalar(Maybe[Int] $min, Maybe[Int] $max) : Ref

    The scalar method returns a random scalar reference.

    scalar example #1

        # given: synopsis
      
        $t->scalar;

 scalar_object

      scalar_object(Maybe[Int] $min, Maybe[Int] $max) : Object

    The scalar_object method returns a random scalar object.

    scalar_object example #1

        # given: synopsis
      
        $t->scalar_object;

 string

      string(Maybe[Int] $min, Maybe[Int] $max) : Str

    The string method returns a random string.

    string example #1

        # given: synopsis
      
        $t->string;

 string_object

      string_object(Maybe[Int] $min, Maybe[Int] $max) : Object

    The string_object method returns a random string object.

    string_object example #1

        # given: synopsis
      
        $t->string_object;

 test

      test(Str $name, Int $cycles, ArrayRef[ArrayRef] $spec, CodeRef $callback) : Undef

    The test method generates subtests using "subtest" in Test::More,
    optionally generating and passing random values to each iteration as
    well as a Data::Object::Try object for easy execution of callbacks and
    interception of exceptions. This callback expected should have the
    signature ($tryable, @arguments) where @arguments gets assigned the
    generated values in the order specified. The callback must return the
    $tryable object, which is called for you automatically, executing the
    subtest logic you've implemented.

    test example #1

        # given: synopsis
      
        # use Test::More;
      
        sub is_an_adult {
          my ($age) = @_;
      
          $age >= 18;
        }
      
        $t->test('is_an_adult', 100, [['number', [10, 30]]], sub {
          my ($tryable, $age) = @_;
      
          $tryable->call(sub {
            if ($age >= 18) {
              ok is_an_adult($age),
                "age is $age, is an adult";
            }
            else {
              ok !is_an_adult($age),
                "age is $age, is not an adult";
            }
          });
      
          $tryable
        });

 time

      time(Maybe[Str] $min, Maybe[Str] $max) : Str

    The time method returns a random time.

    time example #1

        # given: synopsis
      
        $t->time;

 undef

      undef() : Undef

    The undef method returns the undefined value.

    undef example #1

        # given: synopsis
      
        $t->undef;

 undef_object

      undef_object() : Object

    The undef_object method returns the undefined value as an object.

    undef_object example #1

        # given: synopsis
      
        $t->undef_object;

 word

      word() : Str

    The word method returns a random word.

    word example #1

        # given: synopsis
      
        $t->word;

 words

      words(Maybe[Int] $min, Maybe[Int] $max) : Str

    The words method returns random words.

    words example #1

        # given: synopsis
      
        $t->words;

AUTHOR

    Al Newkirk, awncorp@cpan.org

LICENSE

    Copyright (C) 2011-2019, Al Newkirk, et al.

    This is free software; you can redistribute it and/or modify it under
    the terms of the The Apache License, Version 2.0, as elucidated in the
    "license file"
    <https://github.com/iamalnewkirk/foobar/blob/master/LICENSE>.

PROJECT

    Wiki <https://github.com/iamalnewkirk/foobar/wiki>

    Project <https://github.com/iamalnewkirk/foobar>

    Initiatives <https://github.com/iamalnewkirk/foobar/projects>

    Milestones <https://github.com/iamalnewkirk/foobar/milestones>

    Contributing
    <https://github.com/iamalnewkirk/foobar/blob/master/CONTRIBUTE.md>

    Issues <https://github.com/iamalnewkirk/foobar/issues>



CHANGES
Changelog for Test-Ability

0.05 2020-04-19T13:11:48
 - Added release updates
 - Add missing dependecny

0.04 2020-04-19T01:51:13
 - Built release version 0.04
 - Added release updates
 - Fix weird specref modification

0.03 2020-04-09T17:29:52
 - Built release version 0.03
 - Added release updates
 - Fix number generator test failing on zero

0.02 2020-04-09T04:01:11
 - Built release version 0.02
 - Added release updates
 - Enhance documentation

0.01 2020-04-09T03:10:16
 - Built release version 0.01
 - Added release updates
 - Add META files
 - Update documentation
 - Initial implementation
 - Initial commit


lib/Test/Ability.pm
package Test::Ability;

use 5.014;

use strict;
use warnings;

use registry;
use routines;

use Data::Object::Class;
use Data::Object::ClassHas;
use Data::Object::Try;

use Clone ();
use Data::Random ();
use Test::More ();

with 'Data::Object::Role::Buildable';
with 'Data::Object::Role::Stashable';

# VERSION

# ATTRIBUTES

has 'arguments' => (
  is => 'ro',
  isa => 'ArrayRef',
  opt => 1,
);

has 'invocant' => (
  is => 'ro',
  isa => 'Object',
  opt => 1,
);

# BUILD

method build_arg($data) {
  {
    invocant => $data
  }
}

method build_self($data) {
  $self->stash(array => $self->can('array'));
  $self->stash(array_object => $self->can('array_object'));
  $self->stash(code => $self->can('code'));
  $self->stash(code_object => $self->can('code_object'));
  $self->stash(date => $self->can('date'));
  $self->stash(datetime => $self->can('datetime'));
  $self->stash(hash => $self->can('hash'));
  $self->stash(hash_object => $self->can('hash_object'));
  $self->stash(instance => $self->can('instance'));
  $self->stash(number => $self->can('number'));
  $self->stash(number_object => $self->can('number_object'));
  $self->stash(object => $self->can('object'));
  $self->stash(regexp => $self->can('regexp'));
  $self->stash(regexp_object => $self->can('regexp_object'));
  $self->stash(scalar => $self->can('scalar'));
  $self->stash(scalar_object => $self->can('scalar_object'));
  $self->stash(string => $self->can('string'));
  $self->stash(string_object => $self->can('string_object'));
  $self->stash(time => $self->can('time'));
  $self->stash(undef => $self->can('undef'));
  $self->stash(undef_object => $self->can('undef_object'));
  $self->stash(word => $self->can('word'));
  $self->stash(words => $self->can('words'));

  return $self;
}

# METHODS

method array(Maybe[Int] $min = 5, Maybe[Int] $max = 10) {
  my $data = [
    Data::Random::rand_chars(
      set => 'all', min => $min, max => $max
    )
  ];

  return $data;
}

method array_object(Maybe[Int] $min = 5, Maybe[Int] $max = 10) {
  my $data = $self->array;
  my $word = $self->word;

  return $self->instance($data, ucfirst($word));
}

method choose(ArrayRef[ArrayRef] $spec = [[]]) {
  my $item = $spec->[rand($#$spec + 1)];
  my $data = $self->dispatch($item);

  return $data;
}

method code(Maybe[Int] @args) {
  my $data = sub {
    ($self->words(@args), $self->number(@args))[rand(2)]
  };

  return $data;
}

method code_object(Maybe[Int] $min, Maybe[Int] $max) {
  my $data = $self->code;
  my $word = $self->word;

  return $self->instance($data, ucfirst($word));
}

method date(Maybe[Str] $min, Maybe[Str] $max) {
  my $data = Data::Random::rand_date(
    min => $min,
    max => $max
  );

  return $data;
}

method datetime(Maybe[Str] $min, Maybe[Str] $max) {
  my $data = Data::Random::rand_datetime(
    min => $min,
    max => $max
  );

  return $data;
}

method dispatch(ArrayRef $item = []) {
  my ($method, $arguments) = @$item;

  if (!$method) {
    return undef;
  }
  else {
    if (my $callback = $self->stash($method)) {
      return $callback->($self, @$arguments);
    }
    else {
      return $self->$method(@$arguments);
    }
  }
}

method hash(Maybe[Int] $min, Maybe[Int] $max) {
  my $data = $self->array($min, $max);

  $data = {
    @$data % 2 ? (@$data, $$data[rand(@$data)]) : @$data
  };

  return $data;
}

method hash_object(Maybe[Int] $min, Maybe[Int] $max) {
  my $data = $self->hash;
  my $word = $self->word;

  return $self->instance($data, ucfirst($word));
}

method instance(Any $value, Str $class) {

  return bless $value, $class;
}

method maybe(ArrayRef[ArrayRef] $spec = [[]]) {
  my $data = ($self->choose($spec), undef)[rand(2)];

  return $data;
}

method number(Maybe[Int] $min = 0, Maybe[Int] $max = 1_000) {
  $min = 0 if !$min || $min > $max;

  my $data = $min + int rand($max - $min);

  return $data;
}

method number_object(Maybe[Int] $min = 0, Maybe[Int] $max = 1_000) {
  my $data = $self->number($min, $max);
  my $word = $self->word;

  return $self->instance(\$data, ucfirst($word));
}

method object() {
  my @objects = qw(
    array_object
    code_object
    hash_object
    number_object
    regexp_object
    scalar_object
    string_object
    undef_object
  );

  my $method = $objects[rand(@objects)];

  return $self->$method;
}

method regexp(Str $exp = '.*') {
  my $word = $self->word;
  my $data = ref($exp) ? $exp : qr/$exp/;

  return $data;
}

method regexp_object(Str $exp = '.*') {
  my $data = $self->regexp($exp);
  my $word = $self->word;

  return $self->instance(\$data, ucfirst($word));
}

method scalar(Maybe[Int] @args) {
  my $data = ($self->words(@args), $self->number(@args))[rand(2)];

  return \$data;
}

method scalar_object(Maybe[Int] $min, Maybe[Int] $max) {
  my $data = $self->scalar;
  my $word = $self->word;

  return $self->instance($data, ucfirst($word));
}

method string(Maybe[Int] $min, Maybe[Int] $max) {
  my $data = $self->words($min, $max);

  return $data;
}

method string_object(Maybe[Int] $min, Maybe[Int] $max) {
  my $data = $self->string;
  my $word = $self->word;

  return $self->instance(\$data, ucfirst($word));
}

method test(Str $name, Int $cycles, ArrayRef[ArrayRef] $spec, CodeRef $callback) {
  my @defaults;

  if ($self->invocant) {
    push @defaults, invocant => $self->invocant;
  }

  if ($self->arguments) {
    push @defaults, arguments => $self->arguments;
  }

  Test::More::subtest($name, sub {
    for my $index (1..$cycles) {
      Test::More::subtest("$name ($index of $cycles)", sub {
        $callback->(Data::Object::Try->new(@defaults),
          map $self->dispatch($_), map Clone::clone($_), @$spec)->result
      });
    }
  });

  return;
}

method time(Maybe[Str] $min, Maybe[Str] $max) {
  my $data = Data::Random::rand_time(
    min => $min,
    max => $max
  );

  return $data;
}

method undef() {

  return undef;
}

method undef_object() {
  my $data = undef;
  my $word = $self->word;

  return $self->instance(\$data, ucfirst($word));
}

method word() {
  my $data = Data::Random::rand_words();

  return $data->[0];
}

method words(Maybe[Int] $min = 2, Maybe[Int] $max = 10) {
  my $data = Data::Random::rand_words(
    min => $min, max => $max
  );

  return join ' ', @$data;
}

1;


.gen/lib/Test/Ability.pm
package Test::Ability;

use 5.014;

use strict;
use warnings;

use registry;
use routines;

use Data::Object::Class;
use Data::Object::ClassHas;

extends 'Data::Object::Try';

with 'Data::Object::Role::Buildable';
with 'Data::Object::Role::Stashable';

# VERSION

# ATTRIBUTES

has 'arguments' => (
  is => 'ro',
  isa => 'ArrayRef',
  opt => 1,
);

has 'invocant' => (
  is => 'ro',
  isa => 'Object',
  opt => 1,
);

# METHODS

method array() {
  # do something ...

  return $self;
}

method array_object() {
  # do something ...

  return $self;
}

method choose() {
  # do something ...

  return $self;
}

method code() {
  # do something ...

  return $self;
}

method code_object() {
  # do something ...

  return $self;
}

method date() {
  # do something ...

  return $self;
}

method datetime() {
  # do something ...

  return $self;
}

method hash() {
  # do something ...

  return $self;
}

method hash_object() {
  # do something ...

  return $self;
}

method maybe() {
  # do something ...

  return $self;
}

method number() {
  # do something ...

  return $self;
}

method number_object() {
  # do something ...

  return $self;
}

method object() {
  # do something ...

  return $self;
}

method regexp() {
  # do something ...

  return $self;
}

method regexp_object() {
  # do something ...

  return $self;
}

method scalar() {
  # do something ...

  return $self;
}

method scalar_object() {
  # do something ...

  return $self;
}

method string() {
  # do something ...

  return $self;
}

method string_object() {
  # do something ...

  return $self;
}

method test() {
  # do something ...

  return $self;
}

method time() {
  # do something ...

  return $self;
}

method undef() {
  # do something ...

  return $self;
}

method undef_object() {
  # do something ...

  return $self;
}

method word() {
  # do something ...

  return $self;
}

method words() {
  # do something ...

  return $self;
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Property-Based Testing for Perl 5",
  "AUTHOR" => "Al Newkirk <awncorp\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Test-Ability",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.014",
  "NAME" => "Test::Ability",
  "PREREQ_PM" => {
    "Clone" => "0.43",
    "Data::Object::Class" => "2.02",
    "Data::Object::ClassHas" => "2.01",
    "Data::Object::Role::Buildable" => "0.03",
    "Data::Object::Role::Stashable" => "2.01",
    "Data::Object::Try" => "2.02",
    "Data::Random" => "0.13",
    "routines" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Clone" => "0.43",
    "Data::Object::Class" => "2.02",
    "Data::Object::ClassHas" => "2.01",
    "Data::Object::Role::Buildable" => "0.03",
    "Data::Object::Role::Stashable" => "2.01",
    "Data::Object::Try" => "2.02",
    "Data::Random" => "0.13",
    "Test::Auto" => "0.10",
    "routines" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.05",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Clone" => "0.43",
  "Data::Object::Class" => "2.02",
  "Data::Object::ClassHas" => "2.01",
  "Data::Object::Role::Buildable" => "0.03",
  "Data::Object::Role::Stashable" => "2.01",
  "Data::Object::Try" => "2.02",
  "Data::Random" => "0.13",
  "Test::Auto" => "0.10",
  "routines" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
